packages/measure/src/measure-renders.tsx

```diff
@@ -55,8 +65,7 @@ export async function measureRender(ui: React.ReactElement, options?: MeasureOpt
     };
 
     const uiToRender = buildUiToRender(ui, handleRender, options?.wrapper);
-    const screen = render(uiToRender);
-
+    screen = render(uiToRender);
```

Comment:
I wonder: doesn't this generate a race condition in which `handleRender` is called earlier than the actual `render` returns?

Pls add logging and check this. Perhaps, for the first render we might need to manually capture `screen.toJSON()` just after calling `render`.

Reply:
Actually this is the part I am so confused. Since `handleRender` defined and called before `screen` how we can capture the json 🤔

Reply:
In addition, Imagine we have a component that updating state on `useEffect`.

`handleRender` will be called but screen will be not initialized yet. So we can not capture the state update even the state update might not effect the UI. For code demonstration;

`const UnnecessaryRendersComponent = () => {
  const [name, setname] = React.useState('');

  React.useEffect(() => {
    setname('Hello');
  }, []);

  return <View />;
};`

Reply:
const screen = render(...);
if (i === 0 && testingLibrary === 'react-native') {
  const firstJson = screen.toJSON();
}

Reply:
regarding `useEffects` with no async delay. We would have to test if they happen before initial `render` returns. It might be so, but that would also by a symptom of a problem, that when component has all the data at hand (i.e. no async operations happen) it is triggering re-render using `useEffect` instead of rendering it once correctly.

Reply:
Yes that works to capture `firstJSON` but I have a question. Since `setname` cause component re-render `React.Profiler` catches the render. Since component is not rendered yet I believe screen is not being initialized. Because I am unable to capture json on `useEffect` caused render

Reply:
Can you add some `console.logs` and post the output with comments on the timing of the events?

Reply:
Sure, Let me put some logs from UnneccessaryRenders component. I've updated the component like this to demonstrate the issue;

`const UnnecessaryComponent = () => {
  const [count, setCount] = React.useState(0);
  const [name, setname] = React.useState('');

  React.useEffect(() => {
    setname('name');
  }, []);

  React.useEffect(() => {
    setCount(1);
  }, [name]);

  const handlePress = () => {
    setCount((c) => c + 1);
  };

  return (
    <View>
      <Pressable accessibilityRole="button" onPress={handlePress}>
        <Text>Action</Text>
      </Pressable>

      <Text>Count: 0</Text>
    </View>
  );
};`


`  console.log
    is screen initialized:  false  render count:  0

      at log (../../packages/measure/src/measure-renders.tsx:53:17)
          at _loop.next (<anonymous>)

  console.log
    is screen initialized:  false  render count:  1

      at log (../../packages/measure/src/measure-renders.tsx:53:17)
          at _loop.next (<anonymous>)

  console.log
    is screen initialized:  false  render count:  2

      at log (../../packages/measure/src/measure-renders.tsx:53:17)
          at _loop.next (<anonymous>)

  console.log
    firstJSON capturing...

      at log (../../packages/measure/src/measure-renders.tsx:76:15)
          at _loop.next (<anonymous>)

  console.log
    is screen initialized:  true  render count:  3`

Reply:
<img width="575" alt="Screenshot 2024-03-19 at 01 08 35" src="https://github.com/callstack/reassure/assets/40958989/80abaa6b-7c98-4d6d-83a5-000043d8e7b9">

Reply:
Awesome thanks, that what I suspected. So without being able to received the rendered `screen` object we cannot access the `toJSON` for these initial renders. This is not a huge deal, as the fact there is more than a single one is an issue we should warn user about. Let's call it "unnecessary initial re-renders". The reasoning behind this is that with the initial data the component should render exactly ones, if it has some `useEffects` this should be used to make async operations, and not sync ones, as the sync ones can be re-organized into a single proper re-render.

Reply:
Next we will do the tree comparison between the final initial render and subsequent deplayed (async) re-renders. These we can simpy call "unnesessary renders" (perhaps you have some better name) and are the original scope of this PR.

Reply:
So actually if `useEffect` contains some sync operations React immediately schedule for re-render operation but async operation React does not schedule a re-render operation it waits for async operation for initially rendering the component. Did I get it correctly 🤔

Reply:
Yes, async operation will probably end after initial render(s). 

To sum up we have to cases when we want to warn users:
1. More than 1 initial render: this means that they put some `useEffect` which triggers immediate re-render. From performance perspective it should be transformed into a single render, as it's is not waiting for addtional data (as it is sync).
2. Two subsequent renders giving the same JSON output. In such case the render did emit the same output and hence could be optimized by improving code, probably be adding memoization, improving hooks subscriptions, etc.

Reply:
I see, I updated the implementation based on that. As a result we have an object holds both situations.

`  redundantRender: {
    initial: number;
    update: number;
  };`
  
`initial` is used for first and `update` is for second one. We know also put the information to output.json as well. And as comparison we do in `Count changes` its apply for `Redundant Render`

Reply:
But I do have a question for comparing subsequentially I've used `pretty-format` package. Do you think it is okay or we should write a function as you said earlier which uses dfs to go through and look for match @mdjastrzebski